test/fs.js

```diff
@@ -0,0 +1,186 @@
+      else cb(err);
+      return;
+    }
+    if (stats.isDirectory()) {
+      fs.readdir(dir, (err, files) => {
+        if (err) {
+          cb(err);
+          return;
+        }
+        files = files.map(file => path.join(dir, file));
+        metasync.each(files, rmdirp, err => {
+          if (err) cb(err);
+          else fs.rmdir(dir, cb);
+        });
+      });
+    } else {
+      fs.unlink(dir, cb);
+    }
+  });
+};
+
+class ToUpperCase extends stream.Transform {
+  constructor(options = {}) {
+    options = Object.assign(options, {
+      decodeStrings: false,
+    });
+    super(options);
+  }
+  _transform(chunk, encoding, callback) {
+    if (encoding !== 'utf8') {
+      this.emit('error', new Error('only utf8'));
+      return callback();
+    }
+    this.push(chunk.toUpperCase());
+    return callback();
+  }
+}
+
+metatests.test('fs / copyDir', test => {
+  createTestDir(srcDir + '1', err => {
+    if (err) test.error(err);
+    copyDir(srcDir + '1', targetDir + '1', {}, err => {
+      if (err) test.error(err);
+      rmdirp(srcDir + '1', err => {
+        if (err) test.error(err);
+        rmdirp(targetDir + '1', err => {
+          if (err) test.error(err);
+          test.end();
+        });
+      });
+    });
+  });
+});
+
+metatests.test('fs / copyDir / filter', test => {
+  createTestDir(srcDir + '2', err => {
+    if (err) test.error(err);
+    copyDir(srcDir + '2', targetDir + '2', { filter: /1.js/ }, err => {
+      if (err) test.error(err);
+      rmdirp(srcDir + '2', err => {
+        if (err) test.error(err);
+        rmdirp(targetDir + '2', err => {
+          if (err) test.error(err);
+          test.end();
+        });
+      });
+    });
+  });
+});
+
+metatests.test('fs / copyDir / dereference', test => {
+  createTestDir(srcDir + '3', err => {
+    if (err) test.error(err);
+    copyDir(srcDir + '3', targetDir + '3', { dereference: true }, err => {
+      if (err) test.error(err);
+      rmdirp(srcDir + '3', err => {
+        if (err) test.error(err);
+        rmdirp(targetDir + '3', err => {
+          if (err) test.error(err);
+          test.end();
+        });
+      });
+    });
+  });
+});
+
+metatests.test('fs / copyDir / transform', test => {
+  createTestDir(srcDir + '4', err => {
+    if (err) test.error(err);
+    copyDir(
+      srcDir + '4',
+      targetDir + '4',
+      {
+        Transform: ToUpperCase,
+        readOptions: { encoding: 'utf-8' },
+        writeOptions: {},
+      },
+      err => {
+        if (err) test.error(err);
+        rmdirp(srcDir + '4', err => {
+          if (err) test.error(err);
+          rmdirp(targetDir + '4', err => {
+            if (err) test.error(err);
+            test.end();
+          });
+        });
+      }
+    );
+  });
+});
+
+metatests.test('fs / copyDir / depth', test => {
+  createTestDir(srcDir + '5', err => {
+    if (err) test.error(err);
+    copyDir(srcDir + '5', targetDir + '5', { depth: 2 }, err => {
```

Comment:
There also should be a test which calls `copyDir` with `depth` which is higher than levels of nested files and vise versa. Also, you should check if all files that are supposed to be copied are indeed were copied and files that should not be copied are not copied.

Reply:
About the last sentence, So, just do something like recursive ```fs.exist```?
Can I change timeout limit somehow? Because if I start 6 tests instead of 5 I get timeout exception, Seems like It needs just some more time to finish deleting, but I am not sure

Reply:
`fs.exists` is deprecated, so it will be better to use `fs.access` instead. Also, do you get a timeout for all tests or just for a particular one? I added several tests and everything seems fine. FYI, timeout is 30 seconds.

Reply:
What do you think about getting array of files from ```fs.readdir``` and then compare it with static array files?

Reply:
I have changed default timeout to 45000 and all tests are passed
Seems like it is hard for my OS to create-copy-delete parallel 6 dirs
also my cooler rotates faster when I check my tests, I hope I won't kill my laptop

Reply:
> Seems like it is hard for my OS to create-copy-delete parallel 6 dirs

That is ok, as long as CI passes there is no need to do something about this.

> What do you think about getting array of files from `fs.readdir` and then compare it with static array files?

Yeah, this is fine. BTW, there is no need in so many files. I think 2 files for each level should be enough.

Reply:
Also one question from me: I need fn which compare two arrays for equality
Does Metarhia have something like this fn, or I should do it in my own?

Reply:
You need to sort `files` from `fs.readdir` and then compare using `test.strictSame`.
```js
test.strictSame(files.sort(), expected)
```

Reply:
And as I understood I should give ```test``` as argument to another fn which checks for equality
something like this
``` javascript
const testTestDir = (current, test, cb) => {
  fs.readdir(current, (err, items) => {
    if (err) {
      cb(err);
      return;
    }
    test.strictSame(items.sort(), files);
    metasync.each(
      items,
      (item, cb) => {
        if (item.includes('dir')) testTestDir(path.join(current, item), test, cb);
        else {
          cb();
          return;
        }
      },
      cb
    );
  });
};
```

Reply:
Same here: why recursion? Almost every recursion can be rewritten to loop and will look much more readable. As for your question: with https://github.com/metarhia/common/pull/251#discussion_r258946046 it can be reduced to this:
```js
copydirTest.test('test 1', test => {
  copyDir(srcFiles[0], dstFiles[0], err => {
    if (err) {
      test.error(err);
      test.end();
      return;
    }
    metasync.each(dstFiles, fs.access, test.cb(() => test.end()));
  });
});
```
Again, this is only one approach to do this and it is can be done differently.

Reply:
Hm, It looks definitely better. But still, If I will do 6 tests for the same ```srcDir``` ```dstDir``` it will cause error because I am trying to create/delete/copy/check the same dir 6 times.

Reply:
How can I call next test only after previous test was finished?

Reply:
> If I will do 6 tests for the same `srcDir` `dstDir` it will cause error because I am trying to create/delete/copy/check the same dir 6 times.

`beforeEach` and `afterEach` are supposed to do exactly that: create directories before the test/clean up after the test. There is no need for a separate directory for each individual test. Tests are running sequentially by default, so there won't be race conditions and you don't have to worry about same directories accessed by several tests at the same time.

> How can I call next test only after previous test was finished?

Tests are executed by [@metarhia/metatests](https://github.com/metarhia/metatests) [`runner`](https://github.com/metarhia/metatests/blob/master/lib/runner.js). Subtests are instances of `ImperativeTest` and behave just like usual `metatests.test`.